### ISSUE TITLE
Infelicities

### ISSUE THREAD

**Issue by jrp2014** [phadej/cabal-fmt#14]
Thanks for putting together this much-needed tool.

A few reformattings that it produces that I don't think are improvements:

* Omission of space after a comparison in the context of a flag:
```
<   if impl(ghc >= 8.4.1) && impl(ghc <= 8.4.3)
<     build-depends: text >= 1.2.3.0 && < 1.3
---
>   if (impl(ghc >=8.4.1) && impl(ghc <=8.4.3))
>     build-depends: text >=1.2.3.0 && <1.3
```

* Comments in the middle of lists get pushed to the end of the list, which separates them from their object

* It would be good to wrap overlong lines and get rid of extra whitespace:
```
>     , process-extras    >=0.3.0    && <0.3.4 || >=0.4.1.3 && <0.5 || >=0.7.1 && <0.8
```

* Some fields are aligned, some not
```
>   type:               exitcode-stdio-1.0
>   hs-source-dirs:     test/
>   main-is:            Main.hs
>   other-modules:
>     Bugs.Tests
>     Compiler.Tests
```

* `ghc-options` fields are turned from vertical lists to paragraphs

**Comment by phadej** [comment]
- the first point is true, I don't have custom code for pretty printing conditionals (also extra parentheses is not nice)
- That's known issue and won't be fixed. https://github.com/phadej/cabal-fmt/blob/2378d305790f0ffb6546316128bdd3d2ecd66768/src/CabalFmt/Comments.hs#L41
  The proper style is instead of single field with intra comment

  ```cabal
  build-depends:
    , base         ^>=4.10
      -- some comment
    , transformers ^>= 0.6.0.1
  ```

  split it into two

  ```cabal
  build-depends:
    , base ^>= 4.10

  -- some comment
  build-depends:
    , transformers ^>=0.6.0.1
  ```

The last three are style preferences. I do like all three like they are now.

- I'm honestly waiting for Cabal-3.0 to be supported by Hackage, so one 
  can write

  ```cabal
  build-depends:
    , process-extras ^>= { 0.3.0, 0.4.1.3, 0.7.1 }
  ```

- Multiline fields are not aligned. I simply don't like the

  ```
  type:               exitcode-stdio-1.0
  hs-source-dirs:     test/
  main-is:            Main.hs
  other-modules:      Bugs.Tests
                      Compiler.Tests 
  ```

  style.

- For `ghc-options` too, I prefer

  ```cabal
  ghc-options: -Wall -Wextra
  ```
  
  and not
  
  ```cabal
  ghc-options:
    -Wall
    -Wextra
  ```

---

So in summary. First one is something I'd like to see improved (and probably will do that eventually myself).
The rest, I don't like them. You have to write a patch and make it configurable (leaving current behaviour as default).

**Comment by jrp2014** [comment]
Thanks @phadej.  Some of this "depends".  For example, `ghc-options` is indeed better on one line, when there are just a few options, but when there are many, it's hard to read.  It's be nice to have then in alphabetical order, but some of them are order-sensitive (eg, -Werror).

I think that the other multi-line fields may have similar dichotomies.

On the comments, the reality is that things get commented out temporarily until something else is fixed, so adding extra `build-depends`, or whatever, is not going to happen.

Anyway.  I appreciate the work that has gone into this.  There are just a couple of points that make it harder for this to become a canonical app.

**Comment by phadej** [comment]
I simply don't have any bandwidth to think about stuff in `cabal-fmt` which doesn't bother me. (first point does, but not enough I would do something about it).

I made an external tool (i.e. `cabal-fmt`) instead of improving `cabal format` because I wanted something simple (and different than current `cabal format`). If `cabal-fmt` works for others: that is great, otherwise and especially in "depends" cases, I'm sorry, I cannot help you here.
